nncf/torch/dynamic_graph/io_handling.py

```diff
@@ -124,25 +249,33 @@ def wrap_inputs(self, model_args, model_kwargs):
 
             if param_name not in bound_model_params.arguments:
                 nncf_logger.warning(
-                    "A call to a compressed model's forward occurred without one of the params"
-                    "specified in input_infos! Input compression may be incorrect. Trying to recover "
-                    "by wrapping the default value for the parameter."
+                    "A call to a compressed model's forward occurred without one of the arguments "
+                    "specified at the compressed model creation stage! Input compression may be incorrect. "
+                    "Trying to recover by wrapping the default value for the argument."
                 )
                 bound_model_params.apply_defaults()
 
             potential_tensor = bound_model_params.arguments[param_name]
             if potential_tensor is not None:
-                bound_model_params.arguments[param_name] = nncf_model_input(bound_model_params.arguments[param_name])
+                bound_model_params.arguments[param_name] = nncf_model_input(potential_tensor)
             else:
                 # Default was None - cannot wrap as-is. Will wrap a dummy tensor as specified in
-                # input infos - will preserve the call order of nncf_model_input nodes,
+                # input info - will preserve the call order of nncf_model_input nodes,
                 # and the post-hooks for the input node will execute. The result won't go anywhere, though.
                 nncf_logger.info(f"Wrapping a dummy tensor for input {param_name}")
-                info_for_missing_input = self._fwd_params_to_input_infos_odict[param_name]
                 device = "cuda"
                 if self._module_ref_for_device is not None:
                     device = get_model_device(self._module_ref_for_device)
-                dummy_tensor = create_mock_tensor(info_for_missing_input, device)
-                _ = nncf_model_input(dummy_tensor)
+                dummy_input_copy = dummy_input.clone().to(device)
+                _ = nncf_model_input(dummy_input_copy)
 
         return bound_model_params.args, bound_model_params.kwargs
+
+
+class HasDataloader(Protocol):
```

Comment:
Could you please explain the necessity of this class?

Reply:
I need it in `ExactInputsInfo.from_nncf_config_dataloaders` in the line `extra_struct: HasDataloader` to tell the future type checker or developer about what is expected from an extra struct to be considered for providing a dataloader. We usually do this by subclassing, but it doesn't look like that we need an extra explicit class in the inheritance hierarchy of `NNCFExtraConfigStruct`s for this purpose alone. Looks like `Protocol`s solve the problem of having to specify required methods for purposes of typing without explicit inheritance. Actually this is the first time I encountered a place where one would need a `Protocol`, and I'm not entirely sure that this is the right way to use it.

I noticed that in the previous version the `EXTRA_STRUCTS_WITH_DATALOADER` was typed as `List[Type[NNCFExtraConfigStruct]]`. I changed it to `List[Type[HasDataloader]]`, maybe it looks better now. I do not yet know how to type a list of classes which both should subclass from one class (`NNCFExtraConfigStruct` in our case) and implement a protocol defined by another class (`HasDataloader`), perhaps you know of a solution.

Reply:
It is not clear for me why this approach doesn't work
```python

if isinstance(extra_struct, (QuantizationRangeInitArgs, BNAdaptationInitArgs)):
    # We know that `data_loader` property exist

# INSTEAD OF

if extra_struct.__class__ in EXTRA_STRUCTS_WITH_DATALOADERS:
    extra_struct: HasDataloader
```

Reply:
Could you please check if PyCharm still shows an error for this exact code?
```python
if isinstance(extra_struct, (QuantizationRangeInitArgs, BNAdaptationInitArgs)):
    ...
```
Visual Studio Code doesn't show any warnings for the provided code. Nevertheless, I don't think adding a class to resolve IDE issues is a good idea.

Reply:
PyCharm doesn't show warnings for your variant too, but I need a list of compatible types as a "constant" in one place. It's not that pycharm has an issue, it's more likely that the static type deduction system in Python does not allow to handle this case reliably.